addons/l10n_in_edi_ewaybill/models/account_edi_format.py

```diff
@@ -0,0 +1,613 @@
+        }
+        return self._l10n_in_edi_connect_to_server(company, url_path="/iap/l10n_in_edi/1/get_ewaybill_by_irn", params=params)
+
+    #=============================== E-waybill API methods ===================================
+
+    @api.model
+    def _l10n_in_edi_ewaybill_no_config_response(self):
+        return {"error": [{
+            "code": "000",
+            "message": _(
+                "A username and password still needs to be set or it's wrong for the E-waybill(IN). "
+                "It needs to be added and verify in the Settings."
+            )}
+        ]}
+
+    @api.model
+    def _l10n_in_edi_ewaybill_check_authentication(self, company):
+        sudo_company = company.sudo()
+        if sudo_company.l10n_in_edi_ewaybill_username and sudo_company._l10n_in_edi_ewaybill_token_is_valid():
+            return True
+        elif sudo_company.l10n_in_edi_ewaybill_username and sudo_company.l10n_in_edi_ewaybill_password:
+            authenticate_response = self._l10n_in_edi_ewaybill_authenticate(company)
+            if not authenticate_response.get("error"):
+                return True
+        return False
+
+    @api.model
+    def _l10n_in_edi_ewaybill_connect_to_server(self, company, url_path, params):
+        user_token = self.env["iap.account"].get("l10n_in_edi")
+        params.update({
+            "account_token": user_token.account_token,
+            "dbuuid": self.env["ir.config_parameter"].sudo().get_param("database.uuid"),
+            "username": company.sudo().l10n_in_edi_ewaybill_username,
+            "gstin": company.vat,
+        })
+        if company.sudo().l10n_in_edi_production_env:
+            default_endpoint = DEFAULT_IAP_ENDPOINT
+        else:
+            default_endpoint = DEFAULT_IAP_TEST_ENDPOINT
+        endpoint = self.env["ir.config_parameter"].sudo().get_param("l10n_in_edi_ewaybill.endpoint", default_endpoint)
+        url = "%s%s" % (endpoint, url_path)
+        try:
+            return jsonrpc(url, params=params, timeout=25)
+        except AccessError as e:
+            _logger.warning("Connection error: %s", e.args[0])
+            return {
+                "error": [{
+                    "code": "404",
+                    "message": _("Unable to connect to the E-WayBill service."
+                        "The web service may be temporary down. Please try again in a moment.")
+                }]
+            }
+
+    @api.model
+    def _l10n_in_edi_ewaybill_authenticate(self, company):
+        params = {"password": company.sudo().l10n_in_edi_ewaybill_password}
+        response = self._l10n_in_edi_ewaybill_connect_to_server(
+            company, url_path="/iap/l10n_in_edi_ewaybill/1/authenticate", params=params
+        )
+        if response and response.get("status_cd") == "1":
+            company.sudo().l10n_in_edi_ewaybill_auth_validity = fields.Datetime.now() + timedelta(
+                hours=6, minutes=00, seconds=00)
+            self.env.cr.commit()
```

Comment:
We need to be careful with the intermediate commits...  Do we need the refresh_time?  Can not we just handle the error at that point?  (We might already do so)

Reply:
yeah that's concerning. Why is there a commit() here?

Reply:
I also do not know why it's here. I will remove it